src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -136,10 +360,56 @@ Object invoke(Object[] args) {
                 layout.dump(abi.arch, argsPtr, System.err);
             }
 
-            return function.returnLayout().isEmpty()
-                    ? null
-                    : jdk.internal.foreign.abi.BindingInterpreter.box(callingSequence.returnBindings(),
-                    s -> argsPtr.addOffset(layout.retOffset(s))); // buffers are leaked
+            if (returnBindings.length == 0) {
+                return null;
+            } else if (returnBindings.length == 1) {
+                Binding.Move move = returnBindings[0];
+                VMStorage storage = move.storage();
+                return SharedUtils.read(argsPtr.addOffset(layout.retOffset(storage)), move.type());
+            } else { // length > 1
+                Object[] returns = new Object[returnBindings.length];
+                for (int i = 0; i < returnBindings.length; i++) {
+                    Binding.Move move = returnBindings[i];
+                    VMStorage storage = move.storage();
+                    returns[i] = SharedUtils.read(argsPtr.addOffset(layout.retOffset(storage)), move.type());
+                }
+                return returns;
+            }
+        } finally {
+            if (stackArgsSeg != null) {
+                stackArgsSeg.close();
+            }
+        }
+    }
+
+    Object invokeInterpBindings(Object[] args, MethodHandle leaf,
+                                Map<VMStorage, Integer> argIndexMap,
+                                Map<VMStorage, Integer> retIndexMap) throws Throwable {
+        List<MemorySegment> tempBuffers = new ArrayList<>();
```

Comment:
Should this also be expressed in terms of NativeScope?

Reply:
Done.